### ISSUE TITLE
Text format revisions

### ISSUE THREAD

**Issue by tlively** [WebAssembly/spec#884]
At the Oct. 2 CG meeting, there was unanimous consent to consider making changes to the text format to improve the consistency of instruction names and remove special characters. Specifically, interest has been expressed in removing slashes from all type conversion (wrap, trunc, extend, convert, demote, promote, and reinterpret) instructions and unifying the formats of memory and global instructions.

To minimize ecosystem disruption, we will plan to make all changes we agree on at once and update as many tools as possible at roughly the same time. We would like to resolve this issue quickly to avoid holding up the standardization process.

What final changes do we want to make?

**Comment by binji** [comment]
OK, so then the full list would be like this, right?

```
;; core instructions
get_local -> local.get
set_local -> local.set
tee_local -> local.tee
get_global -> global.get
set_global -> global.set
i32.wrap/i64 -> i32.wrap_i64
i32.trunc_s/f32 -> i32.trunc_f32_s
i32.trunc_u/f32 -> i32.trunc_f32_u
i32.trunc_s/f64 -> i32.trunc_f64_s
i32.trunc_u/f64 -> i32.trunc_f64_u
i64.extend_s/i32 -> i64.extend_i32_s
i64.extend_u/i32 -> i64.extend_i32_u
i64.trunc_s/f32 -> i64.trunc_f32_s
i64.trunc_u/f32 -> i64.trunc_f32_u
i64.trunc_s/f64 -> i64.trunc_f64_s
i64.trunc_u/f64 -> i64.trunc_f64_u
f32.convert_s/i32 -> f32.convert_i32_s
f32.convert_u/i32 -> f32.convert_i32_u
f32.convert_s/i64 -> f32.convert_i64_s
f32.convert_u/i64 -> f32.convert_i64_u
f32.demote/f64 -> f32.demote_f64
f64.convert_s/i32 -> f64.convert_i32_s
f64.convert_u/i32 -> f64.convert_i32_u
f64.convert_s/i64 -> f64.convert_i64_s
f64.convert_u/i64 -> f64.convert_i64_u
f64.promote/f32 -> f64.promote_f32
i32.reinterpret/f32 -> i32.reinterpret_f32
i64.reinterpret/f64 -> i64.reinterpret_f64
f32.reinterpret/i32 -> f32.reinterpret_i32
f64.reinterpret/i64 -> f64.reinterpret_i64

;; saturating-float-to-int instructions
i32.trunc_s:sat/f32 -> i32.trunc_f32_s_sat
i32.trunc_u:sat/f32 -> i32.trunc_f32_u_sat
i32.trunc_s:sat/f64 -> i32.trunc_f64_s_sat
i32.trunc_u:sat/f64 -> i32.trunc_f64_u_sat
i64.trunc_s:sat/f32 -> i64.trunc_f32_s_sat
i64.trunc_u:sat/f32 -> i64.trunc_f32_u_sat
i64.trunc_s:sat/f64 -> i64.trunc_f64_s_sat
i64.trunc_u:sat/f64 -> i64.trunc_f64_u_sat

;; simd instructions
f32x4.convert_s/i32x4 -> f32x4.convert_i32x4_s
f32x4.convert_u/i32x4 -> f32x4.convert_i32x4_u
f64x2.convert_s/i64x2 -> f64x2.convert_i64x2_s
f64x2.convert_u/i64x2 -> f64x2.convert_i64x2_u
i32x4.trunc_s/f32x4:sat -> i32x4.trunc_f32x4_s_sat
i32x4.trunc_u/f32x4:sat -> i32x4.trunc_f32x4_u_sat
i64x2.trunc_s/f64x2:sat -> i64x2.trunc_f64x2_s_sat
i64x2.trunc_u/f64x2:sat -> i64x2.trunc_f64x2_u_sat
```

**Comment by rossberg** [comment]
It would be more consistent to have the `_u`/`_s` go after the `_sat`, for the same reasons mentioned above.

In fact, I suggest establishing the following consistent naming scheme for numeric instruction mnemonics:

* `t.xxx` for all numeric instructions agnostic to signedness
* `t.xxx_u`/`t.xxx_s` for all numeric instructions sensitive to  signedness

where `xxx` describes the specific operation (e.g., `add`). (That already holds for everything but conversions.)

Conversions are then instantiations of this scheme with `xxx` = `yyy_t`, yielding:

* `t.yyy_t` for conversions agnostic to signedness
* `t.yyy_t_u`/`t.yyy_t_s` for conversions sensitive to sign

where `yyy` describes the kind of conversion  (e.g., `trunc` or `trunc_sat`).

**Comment by binji** [comment]
OK, I've updated the comment above to match this.

**Comment by lars-t-hansen** [comment]
Oh, and atomics:

`i32.atomic.load8_u` fits the proposed pattern.  
`i32.atomic.rmw8_u.add` more or less does not.

**Comment by rossberg** [comment]
@lars-t-hansen, I suppose you cannot interpret the type before the dot strictly as the result type of an operator. It's the type that the operator "is defined at" in some looser sense.

**Comment by lars-t-hansen** [comment]
I'm not sure why I wouldn't want to interpret the type strictly as the result type of the operator, though, since that pattern is used very broadly, just not in some cases where it's suddenly about the arguments and the result type is implied (in all cases I can think of the implied result type is i32).  Of course there's a succinctness to the instructions that fall under that exception that would be lost if we applied the general pattern, but I don't know why that is an argument either.

(I'm not just trolling, I actually think the current syntax for comparisons is confusing and counterintuitive.  Even lt_f64 -- without any type-designating prefix -- would have been better than the current f64.lt.)

**Comment by binji** [comment]
I dunno, `i32.lt_f64` looks super weird to me. If we think of the type before the `.` as a result type, many instructions don't make sense, such as `i32.store` or even `memory.size`. Personally I don't think it's that valuable to encode the types in the instruction name anyway, we just need a way to differentiate them.

**Comment by tlively** [comment]
I think of the type before the dot as signifying what type the generic instruction is specialized on. Since all `lt`s return i32, the type before the dot tells me what type the operands are instead, since those still need to be disambiguated.

**Comment by alexcrichton** [comment]
FWIW some possible breakage I think that may want to be considered when doing I believe is:

* Libraries which want to be idiomatic with the current wasm spec will need breaking changes to update their type names to reflect these name changes. For example a parser library in Rust, [`parity-wasm`](https://doc.rs/parity-wasm), would need a breaking change to rename [`Instruction::GetLocal`](https://docs.rs/parity-wasm/0.32.0/parity_wasm/elements/enum.Instruction.html#variant.GetLocal) to `Instruction::LocalGet`. 
* We've got tests in Rust which verify that the correct instruction is generated for a particular function (namely for intrinsics like SIMD and such). These instruction names come from disassembly the wasm binary itself with `wasm2wat` and the instructions to assert for are [embedded in the source code](https://github.com/rust-lang-nursery/stdsimd/blob/c5f9757b0ff2a9fddddbd2f15b9b30816789e20f/coresimd/wasm32/atomic.rs#L76). If `wasm2wat` changes its output, these tests will start to break.
* A final one is one we've talked about in the Rust project historically is the "blog post" problem. This is where (for example) there's blog posts, tutorials, reference materials, etc on the internet today which may reference wasm instructions by the old names. If the standard changes a number of instructions (especially common ones like `get_local`) then all this documentation becomes out of date and needs and update.

To be clear I'm not opposed to changing the instructions to be more consistent, but I wanted to be sure to point out some perhaps more subtle locations that "breakage" can arise in the non-traditional sense. They're all pretty minor and we're certainly more than willing to do the updates on the Rust side of things, but points that may want to be considered!

Also @binji your [comment with an exhaustive list of renames](https://github.com/WebAssembly/spec/issues/884#issuecomment-426433329) is super helpful!

**Comment by binji** [comment]
Right, we've already experienced some of the same issues when changing `grow_memory` to `memory.grow` and `current_memory` -> `memory.size`, and these instructions are much more common.

@lars-t-hansen 	is right about atomics too, these are odd now:

```
i32.atomic.rmw.add 
i64.atomic.rmw.add 
i32.atomic.rmw8_u.add
i32.atomic.rmw16_u.add
i64.atomic.rmw8_u.add
i64.atomic.rmw16_u.add
i64.atomic.rmw32_u.add
...
```

**Comment by rossberg** [comment]
@binji, good point. I propose moving the _u/_s to the end for those as well (which arguably fits existing instructions).

**Comment by binji** [comment]
OK, updated my comment above to include these as well.

**Comment by rossberg** [comment]
Oh, that reminds me of one more thing that I'd like to suggest. In the light of the reference type and GC proposals it would be more consistent if we renamed `anyfunc` to `funcref`, so that it lines up with the likes of anyref, eqref, optref, and possibly other subtypes of ref.

**Comment by aardappel** [comment]
Apologies for the bike-shedding, but I don't see a whole lot of consistency in the use of `.` and `_`. 

So far the most frequent use of `.` has been to separate the type operated upon (input types, not return type) before the `.` from the rest of the instruction. Which to me means operations like `get_local` which are not type specific shouldn't have a `.` in them. And I'm not sure why there are multiple `.` in atomic ops but not in others.

I'd suggest that either `.` has a very intuitive meaning like above, or maybe we should do away with `.` as well, and make `_` the only separator.

In fact, while I am ranting, why does `i32.add` have a type and `get_local` does not? Both participate in completely static dataflow determined by the stack and locals, and both will have completely static types that are verified when bytecode is loaded. If `get_local` loads the wrong type, that is an error, but that is not obvious from the instruction. If `i32.add` receives a wrong operand that is also an error, but here it is more obvious. If instead the instruction was `add` it wouldn't really change anything, beyond knowing which of the 2 operands is wrong.

**Comment by binji** [comment]
I don't think so, we agreed to the change in the CG meeting, and the change to the spec has landed. I think at this point we should update the tools to use the new names.

**Comment by hjorthjort** [comment]
I found what seems to be a trailing error. On page 11 in the PDF (generated Jan 27, 2019), in the bullet list:

> *Conversions*: consume a value of one type and produce a result of another (the source type of the conversion is the one after the “/”).

**Comment by rossberg** [comment]
@hjorthjort, good catch, see #957.
